### ISSUE TITLE
"Reflect" behavior for collecting errors/results and continuing.

### ISSUE THREAD

**Issue by aearly** [caolan/async#942]
Related to #349 #334 #531 .

There have been many requests for a behavior where when an iterator calls back with an error, the error is collected, and execution continues.  There have been a couple attempts at this feature, but the implementations fall short.

In theory, this would be applied to just about any async function, `map`, `series`, `parallel`, `forEachSeries`, etc..  I'd like to support this broadly, but without having to create duplicate implementations for each function.  Something we can wrap existing functions with, without having to modify the implementations.

We also need to settle on what the name is for this type of behavior.  These have been proposed:

*  `mapCollect` / `parallelCollect` / `eachSeriesCollect`
*  `mapContinue` / `parallelContinue` / `eachSeriesContinue`
*  `mapAnyway` / `parallelAnyway` / `eachSeriesAnyway`
*  `mapMaybe` / `parallelMaybe` / `eachSeriesMaybe`
*  `mapResume` / `parallelResume` / `eachSeriesResume`

Lastly -- how should the collected results  be passed to the final callback?

*  `callback(null, {errors: [], results: []])`
*  `callback(null, [Error, results, results, Error, results..])`
*  `callback(null, [{err: null, result: results0}, {err: Error, result: null}, ...])`
*  `callback(null, results, errors)`

**Comment by megawac** [comment]
I would suggest `callback(null, results, errors)` or `callback(null, [Error, obj, obj, Error, etc..])`

**Comment by tommy31** [comment]
Any update of this?

**Comment by aearly** [comment]
Sorry, we've been too busy to make any progress on this.

**Comment by jtwebman** [comment]
I am going to give this a shot as I need it for work :smile:!

**Comment by jtwebman** [comment]
Actually bluebird is switching to a map with a reflect wrapper around each. Maybe that would be easier? So reflect would wrap each of the callbacks and always return successful. The output would then have null as error and a list of reflect objects that with have a function `hasValue()` method that returns true if the reflect `.value` property has a value (ie. didn't error) and if it did error then the `.error` property will have the error.

**Comment by jtwebman** [comment]
We can add a reflectAll function as well that can take a list of functions with callbacks and wrap them all in async.reflect like this:

```javascript
let tasks = [
  function(callback){
    setTimeout(function(){
        callback(null, 'one');
    }, 200);
  },
  function(callback){
    // do some more stuff but error ...
    callback(new Error('bad stuff happened'));
  }
  function(callback){
    setTimeout(function(){
      callback(null, 'two');
    }, 100);
  }
];

async.parallel(async.reflectAll(tasks),
// optional callback
function(err, results){
  // values
  // results[0].value = 'one'
  // results[1].error = Error('bad stuff happened')
  // results[2].value = 'two'
});
```

**Comment by aearly** [comment]
@jtwebman I like your approach.  It means we don't have to create new methods for each flavor of async function, leaving it to the user to use it when continuing is desired.  Also,  +1 for the consistency with Bluebird's direction.

**Comment by danielshir** [comment]
Great idea, although the only thing that irks me is wrapping the tasks themselves.
The code already has major boilerplate that it requires, so how about using a sub object?
I write most of my code inline (e.g. not using a "tasks" var) so I'd prefer it this way.

Say something like : 

```
let tasks = [
  function(callback){
    setTimeout(function(){
        callback(null, 'one');
    }, 200);
  },
  function(callback){
    // do some more stuff but error ...
    callback(new Error('bad stuff happened'));
  }
  function(callback){
    setTimeout(function(){
      callback(null, 'two');
    }, 100);
  }
];

async.reflect().parallel(tasks),
// optional callback
function(err, results){
  // values
  // results[0].value = 'one'
  // results[1].error = Error('bad stuff happened')
  // results[2].value = 'two'
});
```

**Comment by jtwebman** [comment]
@danielshir How is that different then this:

```javascript
let tasks = [
  function(callback){
    setTimeout(function(){
        callback(null, 'one');
    }, 200);
  },
  function(callback){
    // do some more stuff but error ...
    callback(new Error('bad stuff happened'));
  }
  function(callback){
    setTimeout(function(){
      callback(null, 'two');
    }, 100);
  }
];

async.parallel(async.reflectAll(tasks)),
// optional callback
function(err, results){
  // values
  // results[0].value = 'one'
  // results[1].error = Error('bad stuff happened')
  // results[2].value = 'two'
});
```

**Comment by danielshir** [comment]
Well, if that's your writing style then I guess it looks the same.
But, what if you write your functions inline? As I assume most people do.

It's a bit messy for my taste since you're adding another scope (more parentheses) - 

```javascript

async.parallel(async.reflectAll([function(callback){
    setTimeout(function(){
        callback(null, 'one');
    }, 200);
  },
  function(callback){
    // do some more stuff but error ...
    callback(new Error('bad stuff happened'));
  }
  function(callback){
    setTimeout(function(){
      callback(null, 'two');
    }, 100);
}]
),
// optional callback
function(err, results){
  // values
  // results[0].value = 'one'
  // results[1].error = Error('bad stuff happened')
  // results[2].value = 'two'
});
```

**Comment by jtwebman** [comment]
@danielshir Then you can write it like this if you want:

```javascript
async.parallel([function(callback){
    setTimeout(function(){
        callback(null, 'one');
    }, 200);
  },
  function(callback){
    // do some more stuff but error ...
    callback(new Error('bad stuff happened'));
  }
  function(callback){
    setTimeout(function(){
      callback(null, 'two');
    }, 100);
}].map(async.reflect),
// optional callback
function(err, results){
  // values
  // results[0].value = 'one'
  // results[1].error = Error('bad stuff happened')
  // results[2].value = 'two'
});
```

**Comment by danielshir** [comment]
@jtwebman I didn't think of that and that actually looks way better!
The reflect call is right near the finishing callback which will remind the programmer why results look like they do :+1:

**Comment by enrichz** [comment]
I'm very interested in this feature, thanks for your work guys. +1

**Comment by mahlu** [comment]
+1

**Comment by akayami** [comment]
Great feature for me, but I do not see this in 2.0.0-rc.1 - Any hope it will be ready soon ?

**Comment by aearly** [comment]
We're waiting for #1012 to be completed.  No ETA.  It might wait until 2.1.

**Comment by aearly** [comment]
Closed via #1095 !
